runtime/src/legacymm/cpp/Memory.cpp

```diff
@@ -2192,6 +2192,41 @@ void updateHeapRef(ObjHeader** location, const ObjHeader* object) {
   }
 }
 
+template <bool Strict>
+void updateHeapRefInsideArray(const ArrayHeader* array, const ArrayHeader* destinationArray, int index,
```

Comment:
Please correct me if I'm wrong, but it seems that we can simplify the logic. 

It seems that in both cases (`from >= to` and `from < to`) we need to:
1. Copy first `abs(from - to)` elements with old ref releasing.
2. Copy middle elements
3. Copy last `abs(from - to)` elements with ref adding.

The only difference is the direction of array traversing.

What if split each of the loops in `updateHeapRefsInsideOneArray` into 3 loops as stated above? It's just an IMHO, but this approach seems more readable to me than calculating the release/add condition for each element.

Reply:
I also started to do like this, but the order of each of this 3 loops are different, but order is important for coping. And making releaseRef earlier isn't possible, because it calls GC and we can lose these objects before coping is made.

Reply:
I rewrote and clang better optimizes this code, it becomes much faster.